Align Scoped Custom Element Registries with latest specification

annevk opened the pull request:
#### d49fe62f68615a250dc40a95a5a0aad8f6625bab
<pre>
Align Scoped Custom Element Registries with latest specification
<a href="https://bugs.webkit.org/show_bug.cgi?id=295337">https://bugs.webkit.org/show_bug.cgi?id=295337</a>
<a href="https://rdar.apple.com/155438084">rdar://155438084</a>

Reviewed by Ryosuke Niwa.

WebKit&apos;s current implementation is not aligned with the DOM and HTML
standards in these ways:

1. When cloning, children of ShadowRoot nodes can have their registry
   impacted. This is not supposed to happen as that violates
   encapsulation.
2. When cloning, the supplied registry is not treated as a fallback. A
   clone of a node with a non-null registry can end up with a different
   registry. This is not supposed to happen.
3. The global registry is only sometimes confined to its original
   document. As we want consistent behavior, we make it always belong
   to a single document and nodes that use that registry will use a new
   global registry (or null) when moving into another document going
   forward.
4. To ensure that confinement and allow for flexibility, we&apos;re also
   adding exceptions.

One edge case that is not tackled by this PR is that a global registry
element does not become a null registry element when inserted into a
scoped registry document. (Instead it starts using the scoped registry,
which is wrong.)

Canonical link: <a href="https://commits.webkit.org/297779@main">https://commits.webkit.org/297779@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/38727bd818518c37348d64f4a21a39a6a71d0546

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/112831 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/32566 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/23044 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/119034 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/63348 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/114793 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/33218 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/41129 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/85869 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/36515 "") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/115778 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/26512 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/101501 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/66175 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/25800 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/19632 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/62793 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/95906 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/19707 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/122256 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/39909 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/29755 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/94732 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/40292 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/97721 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/94471 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24115 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/39597 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/17410 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/36012 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/39795 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/45293 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/39434 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/42768 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/41173 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -630,52 +630,55 @@ bool Element::dispatchSimulatedClick(Event* underlyingEvent, SimulatedClickMouse
+Ref<Element> Element::cloneElementWithoutChildren(Document& document, CustomElementRegistry* fallbackRegistry)
 {
-    Ref clone = cloneElementWithoutAttributesAndChildren(document, registry);
+    RefPtr registry = customElementRegistry();
```
Hm... this will ensure the existence of CustomElementRegistry. We probably don't wanna do that.
Also, it seems wasteful to compute the registry just to check for null-ness.
We can instead check `usesNullCustomElementRegistry()` which is a bit flag check.

annevk committed:
Align Scoped Custom Element Registries with latest specification
https://bugs.webkit.org/show_bug.cgi?id=295337
rdar://155438084

Reviewed by Ryosuke Niwa.

WebKit's current implementation is not aligned with the DOM and HTML
standards in these ways:

1. When cloning, children of ShadowRoot nodes can have their registry
   impacted. This is not supposed to happen as that violates
   encapsulation.
2. When cloning, the supplied registry is not treated as a fallback. A
   clone of a node with a non-null registry can end up with a different
   registry. This is not supposed to happen.
3. The global registry is only sometimes confined to its original
   document. As we want consistent behavior, we make it always belong
   to a single document and nodes that use that registry will use a new
   global registry (or null) when moving into another document going
   forward.
4. To ensure that confinement and allow for flexibility, we're also
   adding exceptions.

One edge case that is not tackled by this PR is that a global registry
element does not become a null registry element when inserted into a
scoped registry document. (Instead it starts using the scoped registry,
which is wrong.)

Canonical link: https://commits.webkit.org/297779@main
